File: setup-assistant/DJToolkitSetup/Services/AgentAPI.swift
import Foundation

enum AgentAPIError: LocalizedError {
    case registrationFailed(String)
    case networkError(String)

    var errorDescription: String? {
        switch self {
        case .registrationFailed(let msg): return "Agent registration failed: \(msg)"
        case .networkError(let msg): return "Network error: \(msg)"
        }
    }
}

struct AgentRegisterResponse: Decodable {
    let agent_id: String
    let api_key: String
    let message: String?
}

struct AgentAPI {
    let cloudURL: String

    /// Register a new agent using the JWT from OAuth.
    func registerAgent(jwt: String, machineName: String) async throws -> AgentRegisterResponse {
        guard let url = URL(string: "\(cloudURL)/api/agents/register") else {
            throw AgentAPIError.networkError("Invalid cloud URL")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(jwt)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ["machine_name": machineName]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw AgentAPIError.networkError("Invalid response")
        }

        switch httpResponse.statusCode {
        case 201:
            return try JSONDecoder().decode(AgentRegisterResponse.self, from: data)
        case 401:
            throw AgentAPIError.registrationFailed("Authentication expired. Please sign in again.")
        case 429:
            throw AgentAPIError.registrationFailed("Too many registration attempts. Please wait and try again.")
        default:
            let body = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw AgentAPIError.registrationFailed("Server returned \(httpResponse.statusCode): \(body)")
        }
    }
}

File: setup-assistant-windows/DJToolkitSetup/Services/AgentAPI.cs
using System;
using System.Net.Http;
using System.Net.Http.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace DJToolkitSetup.Services;

public record AgentRegistration(
    [property: JsonPropertyName("api_key")] string ApiKey,
    [property: JsonPropertyName("agent_id")] string AgentId
);

public static class AgentAPI
{
    private static readonly HttpClient _http = new();

    public static async Task<AgentRegistration> RegisterAsync(string cloudUrl, string jwt, string machineName)
    {
        var request = new HttpRequestMessage(HttpMethod.Post, $"{cloudUrl}/api/agents/register")
        {
            Content = JsonContent.Create(new { machine_name = machineName }),
        };
        request.Headers.Authorization = new("Bearer", jwt);

        var response = await _http.SendAsync(request);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadFromJsonAsync<AgentRegistration>()
            ?? throw new Exception("Empty registration response");
    }
}
